lightning/src/chain/chainmonitor.rs

```diff
@@ -82,24 +83,81 @@ where C::Target: chain::Filter,
 	/// descendants of such transactions. It is not necessary to re-fetch the block to obtain
 	/// updated `txdata`.
 	pub fn block_connected(&self, header: &BlockHeader, txdata: &TransactionData, height: u32) {
+		self.process_transactions(header, txdata, |monitor, txdata| {
+			monitor.block_connected(
+				header, txdata, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger)
+		});
+	}
+
+	/// Dispatches to per-channel monitors, which are responsible for updating their on-chain view
+	/// of a channel and reacting accordingly to newly confirmed transactions. For details, see
+	/// [`ChannelMonitor::transactions_confirmed`].
+	///
+	/// Used instead of [`block_connected`] by clients that are notified of transactions rather than
+	/// blocks. May be called before or after [`update_best_block`] for transactions in the
+	/// corresponding block. See [`update_best_block`] for further calling expectations.
+	///
+	/// [`block_connected`]: Self::block_connected
+	/// [`update_best_block`]: Self::update_best_block
+	pub fn transactions_confirmed(&self, header: &BlockHeader, txdata: &TransactionData, height: u32) {
+		self.process_transactions(header, txdata, |monitor, txdata| {
+			monitor.transactions_confirmed(
+				header, txdata, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger)
+		});
+	}
+
+	/// Dispatches to per-channel monitors, which are responsible for updating their on-chain view
+	/// of a channel and reacting accordingly based on the new chain tip. For details, see
+	/// [`ChannelMonitor::update_best_block`].
+	///
+	/// Used instead of [`block_connected`] by clients that are notified of transactions rather than
+	/// blocks. May be called before or after [`transactions_confirmed`] for the corresponding
+	/// block.
+	///
+	/// Must be called after new blocks become available for the most recent block. Intermediary
+	/// blocks, however, may be safely skipped. In the event of a chain re-organization, this only
+	/// needs to be called for the most recent block assuming `transaction_unconfirmed` is called
+	/// for any affected transactions.
+	///
+	/// [`block_connected`]: Self::block_connected
+	/// [`transactions_confirmed`]: Self::transactions_confirmed
+	/// [`transaction_unconfirmed`]: Self::transaction_unconfirmed
+	pub fn update_best_block(&self, header: &BlockHeader, height: u32) {
+		self.process_transactions(header, &[], |monitor, txdata| {
+			// Recursive calls can supply txdata when chain::Filter::register_output returns a
+			// transaction.
+			if txdata.is_empty() {
+				monitor.update_best_block(
+					header, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger)
+			} else {
+				monitor.transactions_confirmed(
```

Comment:
Hmm, right, its tricky - don't really want to have dead code, and it certainly *seems* like if there's no transactions which matched there shouldn't by any new transactions to register, but also good to handle cases. Note that `update_best_block` may be called for blocks where there is a transaction that matches, where we'd have a `transactions_confirmed` call in parallel either before or after the `update_best_block` call. I don't think that changes anything, but its good to keep in mind.

Reply:
I *think* I lean towards dropping it and asserting we never even call register_output if txdata is empty.

Reply:
So essentially in `ChainMonitor::process_transactions`, if `txdata` is empty the result of `process` should also be empty. @ariard How does that sound?

Reply:
Also, the conditional here would turn into an assert that `txdata` is always empty.

Reply:
> So essentially in ChainMonitor::process_transactions, if txdata is empty the result of process should also be empty. 

Yes. That assertion would make sense, I think.

> Also, the conditional here would turn into an assert that txdata is always empty.

Yep, that's also implied.

Reply:
> > So essentially in ChainMonitor::process_transactions, if txdata is empty the result of process should also be empty.
> 
> Yes. That assertion would make sense, I think.
> 

Hmm... so the conditional assertion fails in the following tests:

```term
failures:
    ln::chanmon_update_fail_tests::test_monitor_and_persister_update_fail
    ln::functional_tests::channel_monitor_network_test
    ln::functional_tests::claim_htlc_outputs_single_tx
    ln::functional_tests::htlc_claim_single_commitment_only_a
    ln::functional_tests::test_concurrent_monitor_claim
    ln::functional_tests::test_htlc_on_chain_timeout
    ln::functional_tests::test_update_err_monitor_lockdown
    ln::reorg_tests::test_set_outpoints_partial_claiming
```

In `ln::reorg_tests::test_set_outpoints_partial_claiming`, the failure happens during this call the `connect_blocks`:

https://github.com/jkczyz/rust-lightning/blob/5d1615a605b89a11c630c814279140f4150c5fb3/lightning/src/ln/reorg_tests.rs#L340

```term
TRACE node 1 [lightning::ln::channelmanager : lightning/src/ln/channelmanager.rs, 3498] 0 transactions included in block 6455219ba424fe0bab6c7cf88966c9af4fdbcb85694c84e057410ba5ff7ef5e2 at height 63 provided
TRACE node 1 [lightning::ln::channelmanager : lightning/src/ln/channelmanager.rs, 3523] New best block: 6455219ba424fe0bab6c7cf88966c9af4fdbcb85694c84e057410ba5ff7ef5e2 at height 63
TRACE node 1 [lightning::chain::channelmonitor : lightning/src/chain/channelmonitor.rs, 2103] New best block 243eaf63d91ef68458e5aa28c4d95c631f609f940196581d6e27c272bf37a167 at height 64
TRACE node 1 [lightning::chain::channelmonitor : lightning/src/chain/channelmonitor.rs, 2161] Block 243eaf63d91ef68458e5aa28c4d95c631f609f940196581d6e27c272bf37a167 at height 64 connected with 0 txn matched
INFO  node 1 [lightning::chain::channelmonitor : lightning/src/chain/channelmonitor.rs, 2444] Force-closing channel due to outbound HTLC timeout, HTLC expiry is 61
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 707] Updating claims view at height 64 with 0 matched transactions and 3 claim requests
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 717] Test if outpoint can be aggregated with expiration 64 against 76
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 717] Test if outpoint can be aggregated with expiration 4294967295 against 76
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 717] Test if outpoint can be aggregated with expiration 4294967295 against 76
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 518] Outpoint 77e5943e78d4368cc4089262fa229d44135c653233c56b559180ba5a934204b9:0
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 684] Going to broadcast Holder Transaction e4f4c7a4414e8409a0c6787a2d9254a57e90a914acad12b3c5ed37de6d293002 claiming funding output 0 from 77e5943e78d4368cc4089262fa229d44135c653233c56b559180ba5a934204b9...
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 741] Registering claiming request for 77e5943e78d4368cc4089262fa229d44135c653233c56b559180ba5a934204b9:0
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 745] Broadcast onchain commitment tx
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 518] Outpoint e4f4c7a4414e8409a0c6787a2d9254a57e90a914acad12b3c5ed37de6d293002:0
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 676] Going to broadcast Holder HTLC-Timeout claiming HTLC output 0 from e4f4c7a4414e8409a0c6787a2d9254a57e90a914acad12b3c5ed37de6d293002...
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 741] Registering claiming request for e4f4c7a4414e8409a0c6787a2d9254a57e90a914acad12b3c5ed37de6d293002:0
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 745] Broadcast onchain HTLC-timeout tx
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 518] Outpoint e4f4c7a4414e8409a0c6787a2d9254a57e90a914acad12b3c5ed37de6d293002:1
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 676] Going to broadcast Holder HTLC-Timeout claiming HTLC output 1 from e4f4c7a4414e8409a0c6787a2d9254a57e90a914acad12b3c5ed37de6d293002...
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 741] Registering claiming request for e4f4c7a4414e8409a0c6787a2d9254a57e90a914acad12b3c5ed37de6d293002:1
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 745] Broadcast onchain HTLC-timeout tx
TRACE node 1 [lightning::ln::onchaintx : lightning/src/ln/onchaintx.rs, 859] Bumping 0 candidates
thread 'ln::reorg_tests::test_set_outpoints_partial_claiming' panicked at 'assertion failed: txn_outputs.is_empty()', lightning/src/chain/chainmonitor.rs:147:17
```

Looks like this is because of this line:

https://github.com/jkczyz/rust-lightning/blob/5d1615a605b89a11c630c814279140f4150c5fb3/lightning/src/chain/channelmonitor.rs#L2235

Do we want to watch that output here or wait until we see it on chain?

Reply:
D'oh! Right, I think that's fine to leave? I guess probably leave the code as-is, though I'm somewhat surprised we can't hit it with `ConnectStyle::BestBlockFirst` in those failing tests.

Reply:
> D'oh! Right, I think that's fine to leave?

Could we still replace the conditional code with an assert in `ChainMonitor::update_best_block `? I'm presuming the transaction we are to broadcast couldn't have had its output spent yet -- since we have not yet broadcast the transaction -- so `register_output` shouldn't return anything... which is a little weird because the `WatchOutput` used in the call will have the current block hash set even though the output hasn't made it on chain. 

Let me know if I'm mistaken in the above explanation.

> I guess probably leave the code as-is, though I'm somewhat surprised we can't hit it with `ConnectStyle::BestBlockFirst` in those failing tests.

FWIW, I was running from a rebase prior to modifying `do_connect_block` to use the new `ChainMonitor` interface. So those failures were from `block_connected` hitting those tests.

Reply:
> Could we still replace the conditional code with an assert in ChainMonitor::update_best_block ? I'm presuming the transaction we are to broadcast couldn't have had its output spent ye

For the current case, yes, that's true, but something like this we could easily break and not catch in tests so it worries me. Its not a lot of code and its not hard to see how its correct, even if currently unreachable. If you feel strongly, I'm happy to go with an assertion (or maybe debug assertion?) but for now, meh.

Reply:
Re-wrote the comment and kept as is.

Reply:
I think we can get ride of the `transaction_confirmed` alternative, inserting an assert about `txdata.is_empty()` doesn't break our test coverage which confirms my intuition that this scenario isn't possible per-current-protocol and respecting our current `chain::Filter` API. 

```
diff --git a/lightning/src/chain/chainmonitor.rs b/lightning/src/chain/chainmonitor.rs
index 2f8c7a15..fe7e9f84 100644
--- a/lightning/src/chain/chainmonitor.rs
+++ b/lightning/src/chain/chainmonitor.rs
@@ -126,13 +126,9 @@ where C::Target: chain::Filter,
                self.process_transactions(header, &[], |monitor, txdata| {
                        // While in practice there shouldn't be any recursive calls when given empty txdata,
                        // it's still possible for a chain::Filter implementation to return a transaction.
-                       if txdata.is_empty() {
-                               monitor.update_best_block(
-                                       header, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger)
-                       } else {
-                               monitor.transactions_confirmed(
-                                       header, txdata, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger)
-                       }
+                       assert!(txdata.is_empty());
+                       monitor.update_best_block(
+                               header, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger)
                });
        }
```

Reply:
Sounds good as discussed offline.